doc/rfc/781-lua-system-wide-profiler.md

```diff
@@ -0,0 +1,395 @@
+*/
+LUAMISC_API int luaM_sysprof_set_writer(luam_Sysprof_writer writer);
+
+/*
+** Callback on profiler stopping. Required for correctly cleaning
+** at VM finalization when profiler is still running.
+** Returns zero on success.
+*/
+LUAMISC_API int luaM_sysprof_set_on_stop(luam_Sysprof_on_stop on_stop);
+
+/*
+** Backtracing function for the host stack. Should call `frame_writer` on
+** each frame in the stack in the order from the stack top to the stack
+** bottom. The `frame_writer` function is implemented inside the sysprof
+** and will be passed to the `backtracer` function. If `frame_writer` returns
+** NULL, backtracing should be stopped. If `frame_writer` returns not NULL,
+** the backtracing should be continued if there are frames left.
+*/
+LUAMISC_API int luaM_sysprof_set_backtracer(luam_Sysprof_backtracer backtracer);
+
+/* Start profiler. */
+LUAMISC_API int luaM_sysprof_start(lua_State *L,
+                                   const struct luam_Sysprof_Options *opt);
+
+/* Stop profiler. */
+LUAMISC_API int luaM_sysprof_stop(lua_State *L);
+
+/* Get the profile counters for each VM state. */
+LUAMISC_API int luaM_sysprof_report(struct luam_Sysprof_Counters *counters);
+
+```
+
+The profiler can work in three modes:
+
+```c
+/*
+** DEFAULT mode collects only data for luam_sysprof_counters, which is stored
+** in memory and can be collected with luaM_sysprof_report after profiler
+** stops.
+*/
+#define LUAM_SYSPROF_DEFAULT 0
+/*
+** LEAF mode = DEFAULT + streams samples with only top frames of host and
+** guests stacks in format described in <lj_sysprof.h>
+*/
+#define LUAM_SYSPROF_LEAF 1
+/*
+** CALLGRAPH mode = DEFAULT + streams samples with full callchains of host
+** and guest stacks in format described in <lj_sysprof.h>
+*/
+#define LUAM_SYSPROF_CALLGRAPH 2
+```
+
+
+Available profiler options are stored in the `luam_sysprof_options` structure.
+
+```c
+/* Profiler options. */
+struct luam_Sysprof_Options {
+  /* Profiling mode. */
+  uint8_t mode;
+  /* Sampling interval in msec. */
+  uint64_t interval;
+  /* Custom buffer to write data. */
+  uint8_t *buf;
+  /* The buffer's size. */
+  size_t len;
+  /* Context for the profile writer and final callback. */
+  void *ctx;
+};
+```
+
+As you can see, the writer backend is customizable, and our implementation does not lock it at the C-API level.
+It is required to provide a writer, a backtracer, and an on-stop callback before running the profiler. If one wants to use the default backtracer, he can call `luaM_sysprof_set_backtracer` with `NULL` argument.
+To start the profiler, one needs to fill in the `luam_Sysprof_Options` with desired values and pass it to the `luaM_sysprof_start` function. In the case of a successful start, it returns `PROFILE_SUCCESS`.
+
+
+To stop the profiler, one should call `luaM_sysprof_stop` function, which will return `PROFILE_SUCCESS` in case of successful profiling finalization.
+
+Starting the profiler in Lua is simple:
+
+```lua
```

Comment:
As far as we don't reference this code we can omit line numbers.
Feel free to ignore.